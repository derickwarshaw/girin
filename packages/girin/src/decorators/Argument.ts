import { ArgumentMetadata } from '../metadata/ArgumentMetadata';
import { Generic } from '../metadata/Generic';


export interface ArgumentDecoratorOptions {
  defaultValue?: any;
  description?: string;
}

export function Argument(nameAndType: string, options?: ArgumentDecoratorOptions) {
  const [name, typeString] = nameAndType.split(/\s*:\s*/);

  return function(target: Object, fieldName: string, parameterIndex: number) {
    const definitionClass = target.constructor;
    ArgumentMetadata.create({
      definitionClass,
      fieldName,
      name,
      definedOrder: parameterIndex,
      generic: Generic.of(typeString),
      defaultValue: options && options.defaultValue,
      description: options && options.description,
    });
  }
}