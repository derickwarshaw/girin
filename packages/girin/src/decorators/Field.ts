import { FieldMetadata } from '../metadata/FieldMetadata';
import { Generic } from "../metadata/Generic";
import { globalMetadataStorage } from "../metadata/globalMetadataStorage";

export interface FieldDecoratorOptions {
  description?: string;
  deprecationReason?: string;
}

export function Field(type: string, options?: FieldDecoratorOptions): Function {
  return function(objectType: any, fieldName: string) {
    const definitionClass = objectType.constructor as any;
    FieldMetadata.create({
      fieldName,
      definitionClass,
      description: options && options.description,
      deprecationReason: options && options.deprecationReason,
      generic: Generic.of(type),
      meta: globalMetadataStorage,
    });
  };
}
