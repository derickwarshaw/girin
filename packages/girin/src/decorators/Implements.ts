import { MetadataStorage } from '../metadata/MetadataStorage';
import { ImplementsMetadata } from '../metadata/ImplementsMetadata';


export interface ImplementsDecoratorOptions {
  meta?: MetadataStorage;
}

export function Implements(targetNameOrDefinitionClass: Function | string, options: ImplementsDecoratorOptions = {}) {
  return function(definitionClass: Function & ImplementsDecoratorOptions) {
    const mergedConfig = Object.assign(options, definitionClass, { definitionClass, targetNameOrDefinitionClass });
    ImplementsMetadata.create(mergedConfig);
  };
}