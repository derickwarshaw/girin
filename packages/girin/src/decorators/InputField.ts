
import { InputFieldMetadata } from "../metadata/InputFieldMetadata";
import { MetadataStorage } from "../metadata/MetadataStorage";
import { Generic } from "../metadata/Generic";


export interface InputFieldDecoratorOptions {
  defaultValue?: any;
  description?: string;
  meta?: MetadataStorage;
}

export function InputField(nameAndType: string, options?: InputFieldDecoratorOptions): Function {
  const [ name, typeString ] = nameAndType.split(/\s*:\s*/);

  return function(definitionClass: Function, propertyName: undefined, parameterIndex: number) {
    InputFieldMetadata.create({
      definedOrder: parameterIndex,
      fieldName: name,
      generic: Generic.of(typeString),
      definitionClass: definitionClass,
      description: options && options.description,
      defaultValue: options && options.defaultValue,
      meta: options && options.meta,
    });
  };
}

