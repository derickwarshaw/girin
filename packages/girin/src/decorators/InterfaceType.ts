import { InterfaceTypeMetadata } from '../metadata/InterfaceTypeMetadata';
import { MetadataStorage } from '../metadata/MetadataStorage';
import { GraphQLFieldConfigMap, GraphQLInterfaceTypeConfig } from 'graphql';


export interface InterfaceDecoratorOptions {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLFieldConfigMap<any, any>;

  description?: GraphQLInterfaceTypeConfig<any, any>["description"];
  astNode?: GraphQLInterfaceTypeConfig<any, any>["astNode"];
  resolveType?: GraphQLInterfaceTypeConfig<any, any>["resolveType"];
}

export function InterfaceType(config: InterfaceDecoratorOptions = {}) {
  return function(definitionClass: InterfaceDecoratorOptions & Function) {
    const mergedConfig = Object.assign(config, definitionClass, { name: definitionClass.name, definitionClass });
    InterfaceTypeMetadata.create(mergedConfig);
  };
}
