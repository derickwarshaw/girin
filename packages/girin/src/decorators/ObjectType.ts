import { ObjectTypeMetadata } from '../metadata/ObjectTypeMetadata';
import { MetadataStorage } from '../metadata/MetadataStorage';
import { GraphQLObjectTypeConfig, GraphQLFieldConfigMap } from 'graphql';


export interface ObjectTypeDecoratorOptions {
  name?: string;
  meta?: MetadataStorage;
  fields?: () => GraphQLFieldConfigMap<any, any>;

  interfaces?: GraphQLObjectTypeConfig<any, any>["interfaces"];
  isTypeOf?: GraphQLObjectTypeConfig<any, any>["isTypeOf"];
  description?: GraphQLObjectTypeConfig<any, any>["description"];
  astNode?: GraphQLObjectTypeConfig<any, any>["astNode"];
  extensionASTNodes?: GraphQLObjectTypeConfig<any, any>["extensionASTNodes"];
}

export function ObjectType(options: ObjectTypeDecoratorOptions = {}) {
  return function(definitionClass: ObjectTypeDecoratorOptions & Function) {
    const mergedConfig = Object.assign(options, definitionClass, { name: definitionClass.name, definitionClass });
    ObjectTypeMetadata.create(mergedConfig);
  };
}
