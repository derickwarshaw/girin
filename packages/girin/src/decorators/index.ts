export { Argument } from './Argument';
export { Field } from './Field';
export { Implements } from './Implements';
export { InputField } from './InputField';
export { InputObjectType } from './InputObjectType';
export { ObjectType } from './ObjectType';
