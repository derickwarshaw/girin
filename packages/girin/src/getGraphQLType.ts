import { MetadataStorage } from "./metadata/MetadataStorage";
import { GraphQLType, isType } from 'graphql';

export interface TypeBuilderOptions {
  meta?: MetadataStorage;
}

export function getGraphQLType(definitionClassOrType: Function | GraphQLType, options?: TypeBuilderOptions): any {
  let meta: MetadataStorage;
  if (options && options.meta) {
    meta = options.meta;
  } else {
    meta = MetadataStorage.getMetadataStorage();
  }

  if (isType(definitionClassOrType)) {
    return definitionClassOrType;
  }

  const typeMetadata = meta.getDefinitionMetadata(definitionClassOrType);
  if (!typeMetadata) {
    throw new Error(`Given class ${definitionClassOrType.name} has no corresponding metadata`);
  }
  return typeMetadata.build.typeInstance;
}
