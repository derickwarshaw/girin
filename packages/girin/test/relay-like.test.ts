import { graphql } from "graphql/graphql";
import { printSchema } from "graphql/utilities/schemaPrinter";
import { GraphQLSchema } from "graphql/type/schema";

import { Field } from "../src/decorators/Field";
import { Argument } from "../src/decorators/Argument";
import { ObjectType } from "../src/decorators/ObjectType";
import { Implements } from "../src/decorators/Implements";
import { InterfaceType } from "../src/decorators/InterfaceType";
import { getGraphQLType } from "../src/getGraphQLType";


interface NodeSource {
  id: number;
}

interface GroupSource {
  id: number;
  name: string;
}

interface MemberSource {
  id: number;
  name: string;
  email: string;
}

interface MembershipSource {
  memberId: number;
  groupId: number;
  isActive: boolean;
  since: string;
}

const data = {
  group: [
    { id: 1, name: 'Robots' },
    { id: 2, name: 'Humans' },
  ],
  member: [
    { id: 1, name: 'Key', email: 'k@example.com' },
    { id: 2, name: 'Jonghyun', email: 'j@example.com' },
    { id: 3, name: 'Minho', email: 'm@example.com' },
    { id: 4, name: 'Taemin', email: 't@example.com' }
  ],
  membership: [
    { memberId: 1, groupId: 1, isActive: false, since: '2007-03-12' },
    { memberId: 1, groupId: 2, isActive: true, since: '2001-11-09' },
    { memberId: 2, groupId: 1, isActive: true, since: '2006-02-11' },
    { memberId: 2, groupId: 2, isActive: true, since: '2006-12-11' },
    { memberId: 3, groupId: 1, isActive: true, since: '2011-12-31' },
    { memberId: 4, groupId: 1, isActive: true, since: '2011-12-31' },
  ]
}

@InterfaceType()
abstract class Node {
  constructor(protected source: NodeSource) { }
  @Field('ID!', {
    description: 'The id of the object.',
  })
  public id() {
    const uid = `${this.constructor.name}:${this.source.id}`;
    return Buffer.from(uid, 'utf8').toString('base64');
  }
}


@Implements(Node)
@ObjectType()
class Member extends Node {
  constructor(source: MemberSource) {
    super(source);
    this.email = source.email;
    this.name = source.name;
  }


  @Field('String!') name: string;
  @Field('String!') email: string;
}

@ObjectType()
class MemberEdge {
  constructor(private source: MembershipSource) {
    this.isActive = source.isActive;
    this.since = source.since;
  }

  @Field('Boolean!') isActive: boolean;
  @Field('String!') since: string;

  @Field('Member')
  public node() {
    const memberSource = data.member.find(member => member.id === this.source.memberId);
    if (!memberSource) throw new Error('Given member id has no corresponding member source');
    return new Member(memberSource);
  }
}

@ObjectType()
class MemberConnection {
  protected source: MembershipSource[];
  constructor(source: MembershipSource[]) {
    this.source = source;
  }
  @Field('[MemberEdge]!')
  public edges(): MemberEdge[] {
    return this.source.map(source => new MemberEdge(source));
  }
}

@ObjectType()
@Implements(Node)
class Group extends Node {
  constructor(source: GroupSource) {
    super(source);
    this.name = source.name;
  }

  @Field('String!')
  public name: string;

  @Field('MemberConnection!')
  public members() {
    const source = data.membership.filter(membership => membership.groupId === this.source.id);
    return new MemberConnection(source);
  }
}

@ObjectType()
class Query {
  @Field('Group!')
  public group(
    @Argument('id: Int!') id: number,
  ) {
    const groupSource = data.group.find(group => group.id === id);
    if (!groupSource) throw new Error('Group not found');
    return new Group(groupSource);
  }
}

const schema = new GraphQLSchema({
  query: getGraphQLType(Query),
});


describe('relay-like implementation', () => {
  it('generates schema as expected', () => {
    expect(printSchema(schema)).toMatchSnapshot();
  });

  it('should return expected results when queried', async () => {
    const result = await graphql(schema, `
      fragment GroupFragment on Group {
        id
        name
        members {
          edges {
            isActive
            since
            node {
              id
              name
              email
            }
          }
        }
      }
      query {
        robotsMember: group(id: 1) {
          ...GroupFragment
        }
        humansMember: group(id: 2) {
          ...GroupFragment
        }
        nonFound: group(id: 3) {
          ...GroupFragment
        }
      }
    `);

    expect(JSON.stringify(result, null, '  ')).toMatchSnapshot();
  })
});
