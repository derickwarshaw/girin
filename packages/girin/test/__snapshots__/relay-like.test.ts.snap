// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`relay-like implementation generates schema as expected 1`] = `
"type Group implements Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
  name: String!
  members: MemberConnection!
}

type Member implements Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
  name: String!
  email: String!
}

type MemberConnection {
  edges: [MemberEdge]!
}

type MemberEdge {
  isActive: Boolean!
  since: String!
  node: Member
}

interface Node {
  \\"\\"\\"The id of the object.\\"\\"\\"
  id: ID!
}

type Query {
  group(id: Int!): Group!
}
"
`;

exports[`relay-like implementation should return expected results when queried 1`] = `
"{
  \\"errors\\": [
    {
      \\"message\\": \\"Group not found\\",
      \\"locations\\": [
        {
          \\"line\\": 24,
          \\"column\\": 9
        }
      ],
      \\"path\\": [
        \\"nonFound\\"
      ]
    }
  ],
  \\"data\\": null
}"
`;
